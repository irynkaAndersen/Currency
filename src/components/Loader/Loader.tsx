import React, { FC } from "react";
import { styleNames } from "../../utils/styleNames";
import styles from "./Loader.scss";

const sn = styleNames(styles);

const Loader: FC = () => {
  return (
    <div className={sn("loader")}>
      <div />
      <div />
      <div />
      <div />
    </div>
  );
};

export default Loader;
