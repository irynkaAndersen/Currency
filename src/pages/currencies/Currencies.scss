@import "../../main.scss";

.currencies {
  min-height: 300px;
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  justify-content: center;
}

.error {
  color: $color-red;
}

.sort {
  cursor: pointer;
  background: transparent;
  border: none;
  outline: none;
  font-size: 1.3em;

  &:hover {
    color: $color-red;
  }
}

.time {
  width: 100%;
  text-align: center;
}

.table {
  border-color: #000;
  background: #ababab;
  border-radius: 10px;
  padding: 30px;

  td {
    padding: 10px 20px;
  }

  &_head {
    font-weight: bold;
  }

  &_body {
    td {
      background: #f7fcfc;

      &:nth-child(odd) {
        background: #d0d7da;
      }
    }
  }
}