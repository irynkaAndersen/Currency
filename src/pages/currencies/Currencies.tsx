import React, { FC, useCallback, useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import getBitCoinPrices from "../../actions/currencies.action";
import { getBitcoinPricesSelector } from "../../selectors/currencies.selector";
import { styleNames } from "../../utils/styleNames";
import useSorting from "../../utils/useSorting";
import Loader from "../../components/Loader";
import styles from "./Currencies.scss";

const sn = styleNames(styles);

const Currencies: FC = () => {
  const dispatch = useDispatch();
  const { prices, status, updated } = useSelector(getBitcoinPricesSelector);
  const [sortBy, setSortBy] = useState<{
    key: string;
    direction?: "asc" | "desc";
  }>({ direction: undefined, key: "" });
  const data = useSorting(
    (prices as unknown) as { [x: string]: number | string }[],
    sortBy
  );

  const sortByKey = useCallback(
    (key: string, direction: "asc" | "desc") => () =>
      setSortBy({ key, direction }),
    [setSortBy]
  );

  const showBtn = useCallback(
    (key: string) => {
      let sort: "asc" | "desc" = "asc";
      if (sortBy.key === key) {
        if (sortBy.direction === "asc") sort = "desc";
      }
      return (
        <button
          onClick={sortByKey(key, sort)}
          type="button"
          className={sn("sort")}
        >
          {sort === "asc" ? "↑" : "↓"}
        </button>
      );
    },
    [sortBy]
  );

  const tableData = useMemo(() => {
    const currencies = Object.keys(data);
    return currencies.map((item: string) => {
      // @ts-ignore
      const childKeys = Object.keys(data[item]);
      return (
        <tr key={item}>
          {childKeys.map((tdData: string) => (
            // @ts-ignore
            <td key={tdData}>{data[item][tdData]}</td>
          ))}
        </tr>
      );
    });
  }, [data]);

  const tableHead = useMemo(() => {
    const rows: string[] = Object.keys(data);
    if (rows.length) {
      // @ts-ignore
      const cells = Object.keys(data[rows[0]]);
      return cells.map((item) => (
        <td key={item}>
          {item}
          {showBtn(item)}
        </td>
      ));
    }
    return null;
  }, [data, showBtn]);

  useEffect(() => {
    dispatch(getBitCoinPrices);
    const getDataInterval = setInterval(
      () => dispatch(getBitCoinPrices),
      60000
    );
    return () => clearInterval(getDataInterval);
  }, []);

  const content = useMemo(() => {
    switch (status) {
      case "rejected":
        return <div className={sn("error")}>Service is not responding</div>;
      case "fulfilled":
        return (
          <>
            <table className={sn("table")}>
              <thead className={sn("table_head")}>
                <tr>{tableHead}</tr>
              </thead>
              <tbody className={sn("table_body")}>{tableData}</tbody>
            </table>
            <p className={sn("time")}>Last updated: {updated.toString()}</p>
          </>
        );
      default:
        return <Loader />;
    }
  }, [status, tableData, tableHead, updated]);

  return (
    <div className="wrap">
      <h1>Currencies</h1>
      <div className={sn("currencies")}>{content}</div>
    </div>
  );
};

export default Currencies;
