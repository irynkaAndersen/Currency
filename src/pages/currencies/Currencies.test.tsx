import React from "react";
import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import * as ReactReduxHooks from "react-redux";
import * as enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Loader from "../../components/Loader";
import Currencies from ".";

enzyme.configure({ adapter: new Adapter() });

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn().mockReturnValue({}),
}));

jest.mock("../../selectors/currencies.selector");
jest.useFakeTimers();

describe("Currencies", () => {
  const spy = jest.spyOn(ReactReduxHooks, "useSelector");

  it("should render with data", async () => {
    spy.mockReturnValue({
      prices: {
        EUR: {
          code: "EUR",
          symbol: "&euro;",
          rate: "26,389.9814",
          description: "Euro",
          rate_float: 26389.9814,
        },
        GBP: {
          code: "GBP",
          symbol: "&pound;",
          rate: "23,439.7820",
          description: "British Pound Sterling",
          rate_float: 23439.782,
        },
      },
      status: "fulfilled",
      updated: "2021-01-23T20:42:13.306Z",
    });
    const wrapper = await shallow(<Currencies />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    const table = wrapper.find(".table_body tr");
    expect(table.length).toBe(2);

    expect(wrapper.find(".table_body td").length).toBe(10);
    expect(shallowToJson(wrapper.find(".currencies p")).children).toEqual([
      "Last updated: ",
      "2021-01-23T20:42:13.306Z",
    ]);

    wrapper.find(".table_head td button").at(0).simulate("click");
    expect(wrapper.find(".table_body td").at(0).text()).toBe("EUR");
    wrapper.find(".table_head td button").at(0).simulate("click");
    expect(wrapper.find(".table_body td").at(0).text()).toBe("GBP");
  });

  it("should render with loader", async () => {
    spy.mockReturnValue({
      prices: {},
      status: "pending",
      updated: "2021-01-23T20:42:13.306Z",
    });
    const wrapper = await shallow(<Currencies />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    expect(<Loader />).toBeDefined();
  });

  it("should render with error", async () => {
    spy.mockReturnValue({
      prices: {},
      status: "rejected",
      updated: "2021-01-23T20:42:13.306Z",
    });
    const wrapper = await shallow(<Currencies />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(".error").length).toBe(1);
  });
});
