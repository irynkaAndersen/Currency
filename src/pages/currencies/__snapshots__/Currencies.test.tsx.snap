// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currencies should render with data 1`] = `
<div
  className="wrap"
>
  <h1>
    Currencies
  </h1>
  <div
    className="currencies"
  >
    <table
      className="table"
    >
      <thead
        className="table_head"
      >
        <tr>
          <td
            key="code"
          >
            code
            <button
              className="sort"
              onClick={[Function]}
              type="button"
            >
              ↑
            </button>
          </td>
          <td
            key="symbol"
          >
            symbol
            <button
              className="sort"
              onClick={[Function]}
              type="button"
            >
              ↑
            </button>
          </td>
          <td
            key="rate"
          >
            rate
            <button
              className="sort"
              onClick={[Function]}
              type="button"
            >
              ↑
            </button>
          </td>
          <td
            key="description"
          >
            description
            <button
              className="sort"
              onClick={[Function]}
              type="button"
            >
              ↑
            </button>
          </td>
          <td
            key="rate_float"
          >
            rate_float
            <button
              className="sort"
              onClick={[Function]}
              type="button"
            >
              ↑
            </button>
          </td>
        </tr>
      </thead>
      <tbody
        className="table_body"
      >
        <tr
          key="0"
        >
          <td
            key="code"
          >
            EUR
          </td>
          <td
            key="symbol"
          >
            &euro;
          </td>
          <td
            key="rate"
          >
            26,389.9814
          </td>
          <td
            key="description"
          >
            Euro
          </td>
          <td
            key="rate_float"
          >
            26389.9814
          </td>
        </tr>
        <tr
          key="1"
        >
          <td
            key="code"
          >
            GBP
          </td>
          <td
            key="symbol"
          >
            &pound;
          </td>
          <td
            key="rate"
          >
            23,439.7820
          </td>
          <td
            key="description"
          >
            British Pound Sterling
          </td>
          <td
            key="rate_float"
          >
            23439.782
          </td>
        </tr>
      </tbody>
    </table>
    <p
      className="time"
    >
      Last updated: 
      2021-01-23T20:42:13.306Z
    </p>
  </div>
</div>
`;

exports[`Currencies should render with error 1`] = `
<div
  className="wrap"
>
  <h1>
    Currencies
  </h1>
  <div
    className="currencies"
  >
    <div
      className="error"
    >
      Service is not responding
    </div>
  </div>
</div>
`;

exports[`Currencies should render with loader 1`] = `
<div
  className="wrap"
>
  <h1>
    Currencies
  </h1>
  <div
    className="currencies"
  >
    <Loader />
  </div>
</div>
`;
